text/3488-deprecate-cargo-categories.md

```diff
@@ -0,0 +1,374 @@
+thus being a political decision. If the "recommended keywords" concept is
+implemented primarily based on usage data of existing keywords, this discussion
+can largely be sidestepped, and it gives the Rust community the power to decide
+what keywords should be recommended and presented on the crates.io frontpage.
+
+Finally, having the community decide on the keywords recommendations through
+their usage data reduces the amount of bike-shedding discussions that currently
+exist when categories are decided upon with very little up-front usage data.
+
+
+# Drawbacks
+
+- As mentioned above, the chance for having duplicates is increased since not
+everyone will use the list of recommended keywords. This can be countered to
+some degree with aliases for recommended keywords, but can't be prevented
+completely.
+
+- The main drawback of changing how we categorize crates is the change itself.
+Existing users and crate authors are used to having both concepts available,
+and the existing tooling like IDE autocomplete of categories assumes the same
+thing. It will take a bit of time and educating people to adjust the ecosystem
+to the new concept.
+
+
+# Implementation
+
+The following implementation steps are planned for the crates.io side:
+
+- [ ] Add a publicly visible list of recommended keywords based on existing
+keyword usage on crates.io and the list of existing categories.
+- [ ] Add a publish-time deprecation warning for crates that are published with
+only categories but no keywords.
+- [ ] Remove categories from the crates.io frontend:
+  - Crates inside category pages
+  - Categories list
+  - Category slugs list
+  - Crate details page sidebar
+  - Support for `category:foo` in search bar
+- [ ] Remove categories from the crates.io backend:
+  - `/crates?category=foo` returns an empty list
+  - `/crates/foo` returns an empty list of `categories`
+  - `/category_slugs` returns an empty list
+  - `/categories` returns an empty list
+  - `/categories/foo` returns `404 Not Found`
+  - `/crates/new` stops saving categories for new uploads
+- [ ] Remove categories from that crates.io database (incl. the `ltree`
+extension).
+- [ ] Add a publish-time deprecation warning for all crates that are published
+with categories.
+
+
+# Requirements for recommended keywords
+
+The following basic rules should be followed when promoting keywords to the list
+of recommended keywords.
+
+1. The keyword does not violate the [Code of Conduct](https://www.rust-lang.org/policies/code-of-conduct)
+of the Rust project. (Such keywords are not allowed anyway, but the crates.io
+team is not currently monitoring new additions to the list of keywords and will
+only act on this when users report violations)
+
+2. The keyword is already used by substantial amount of crates from separate
+authors. (Intentionally vague to account for the growth of the package registry)
+
+3. The keyword serves a purpose that is not covered by any other recommended
+keyword.
```

Comment:
Would you mind elaborating a bit on why you removed the bit on crate keywords? My thought process was that if people use the keyword `serde` or similar, it would be redundant to rely on the crates.io team to curate a description for that, and instead it would make sense to simply link to the `serde` crate for that description. We could just exclude these keywords from the list, but I feel like doing something with them is appropriate since they are probably numerous.

Reply:
it felt a bit complicated to me. I'm not sure I understand why we would want to treat keywords that match crate names differently to other keywords. there is also the open question of how to deal with crates that match a keyword but were created after the keyword was already used by other crates.

Reply:
Yeah, that's fair. I mostly thought of it since recommended keywords could have descriptions, and having a description be "they use this crate" felt common enough that it deserved its own category. Probably best left to a future extension, so, I'm fine with that.